lib/auth/server_info.go

```diff
@@ -70,32 +70,59 @@ func (a *Server) ReconcileServerInfos(ctx context.Context) error {
 	}
 }
 
-func (a *Server) setCloudLabelsOnNodes(ctx context.Context, nodes []types.Server) (failedUpdates int, err error) {
+// getServerInfoNames gets the names of ServerInfos that could exist for a
+// node.
+func getServerInfoNames(node types.Server) []string {
+	var names []string
+	if meta := node.GetCloudMetadata(); meta != nil && meta.AWS != nil {
+		names = append(names, types.ServerInfoNameFromAWS(meta.AWS.AccountID, meta.AWS.InstanceID))
+	}
+	// Manually added ServerInfos should override any other ServerInfos.
+	names = append(names, types.ServerInfoNameFromNodeName(node.GetName()))
+	return names
+}
+
+func (a *Server) setLabelsOnNodes(ctx context.Context, nodes []types.Server) (failedUpdates int, err error) {
 	for _, node := range nodes {
-		meta := node.GetCloudMetadata()
-		if meta != nil && meta.AWS != nil {
-			si, err := a.GetServerInfo(ctx, meta.AWS.GetServerInfoName())
+		// Get the server infos that match this node.
+		serverInfoNames := getServerInfoNames(node)
+		serverInfos := make([]types.ServerInfo, 0, len(serverInfoNames))
+		for _, name := range serverInfoNames {
+			si, err := a.GetServerInfo(ctx, name)
 			if err == nil {
-				err := a.updateLabelsOnNode(ctx, node, si)
-				// Didn't find control stream for node, save count for logging.
-				if trace.IsNotFound(err) {
-					failedUpdates++
-				} else if err != nil {
-					return failedUpdates, trace.Wrap(err)
-				}
+				serverInfos = append(serverInfos, si)
 			} else if !trace.IsNotFound(err) {
 				return failedUpdates, trace.Wrap(err)
 			}
 		}
+		if len(serverInfos) == 0 {
+			continue
+		}
+
+		err := a.updateLabelsOnNode(ctx, node, serverInfos)
+		// Didn't find control stream for node, save count for logging.
+		if trace.IsNotFound(err) {
```

Comment:
nit: Use the smaller scope.

```suggestion
		// Didn't find control stream for node, save count for logging.
		if err := a.updateLabelsOnNode(ctx, node, serverInfos); trace.IsNotFound(err) {
```